### PR TITLE
SRU: convert JSON to MARC21

### PR DESCRIPTION
**Pull request by rerowep**
* Adds dojson for json to marc21 transformation.

Co-Authored-by: Peter Weber <peter.weber@rero.ch>

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Cypress tests successful?

### CODE REVIEW

File: rero_ils/modules/documents/dojson/contrib/jsontomarc21/__init__.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# -*- coding: utf-8 -*-
+#
+# RERO ILS
+# Copyright (C) 2019 RERO
```
Comment (iGormilhit) [inline rero_ils/modules/documents/dojson/contrib/jsontomarc21/__init__.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: rero_ils/modules/documents/dojson/contrib/jsontomarc21/__init__.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+"""MARC21 RERO to JSON."""
```
Comment (iGormilhit) [inline rero_ils/modules/documents/dojson/contrib/jsontomarc21/__init__.py:18]:
```suggestion
"""RERO MARC21 to JSON."""
```

File: rero_ils/modules/holdings/api.py [hunk 3]
```diff
@@ -302,34 +303,48 @@ def get_holdings_type_by_holding_pid(cls, holding_pid):
...
+            es_query = es_query.filter(
+                'bool', must_not=[Q('term', _masked=True)])
+        for holding in es_query.scan():
             yield holding.pid
 
     @classmethod
-    def get_holdings_by_document_by_view_code(cls, document_pid, viewcode):
+    def get_holdings_by_document_by_view_code(cls, document_pid, viewcode,
+                                              with_masked=True):
         """Returns holding pids by document and view code."""
```
Comment (iGormilhit) [inline rero_ils/modules/holdings/api.py:334]:
```suggestion
        """Returns holding PIDs by document and view code."""
```

If it's not too much an hassle, change all pid in PID and pids in PIDs in the docstrings. It's easier to read and understand.

File: tests/unit/test_documents_dojson_marc21.py [hunk 4]
```diff
@@ -0,0 +1,338 @@
+# -*- coding: utf-8 -*-
+#
+# RERO ILS
+# Copyright (C) 2019 RERO
```
Comment (iGormilhit) [inline tests/unit/test_documents_dojson_marc21.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: tests/unit/test_documents_dojson_marc21.py [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+"""DOJSON transiformation for ebooks module tests."""
```
Comment (iGormilhit) [inline tests/unit/test_documents_dojson_marc21.py:18]:
```suggestion
"""DOJSON transformation for ebooks module tests."""
```

File: tests/unit/test_documents_dojson_marc21.py [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+from copy import deepcopy
+
+import mock
+from flask_babelex import gettext as _
+
+from rero_ils.modules.documents.dojson.contrib.jsontomarc21 import to_marc21
+
+
+def test_pid_to_marc21(app, marc21_record):
+    """Test pidto marc21 transformation."""
```
Comment (iGormilhit) [inline tests/unit/test_documents_dojson_marc21.py:31]:
```suggestion
    """Test PID to MARC21 transformation."""
```

File: tests/unit/test_documents_dojson_marc21.py [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+    record.update({
+        '__order__': ('leader', '001', '008'),
+        '001': '12345678',
+        '008': '000000|20072020xx#|||||||||||||||||fre|c',
+    })
+    assert result == record
+
+
+def test_title_to_marc21(app, marc21_record):
+    """Test titleto marc21 transformation."""
```
Comment (iGormilhit) [inline tests/unit/test_documents_dojson_marc21.py:73]:
```suggestion
    """Test title to MARC21 transformation."""
```

File: tests/unit/test_documents_dojson_marc21.py [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+            'a': 'Suisse',
+            'b': 'Schweiz. Svizzera : Le guide Michelin 2020'
+        }
+    })
+    assert result == record
+
+
+def test_contribution_to_marc21(app, marc21_record,
+                                mef_record_1, mef_record_2, mef_record_3):
+    """Test contributionto marc21 transformation."""
```
Comment (iGormilhit) [inline tests/unit/test_documents_dojson_marc21.py:164]:
```suggestion
    """Test contribution to MARC21 transformation."""
```

File: tests/unit/test_documents_dojson_marc21.py [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+            'a': 'Congrès ouvrier français',
+            '4': 'aut',
+            '0': ('(idref)03255608X', '(rero)A005462931', '(gnd)050343211')
+        })
+    })
+    assert result == record
+
+
+def test_type_to_marc21(app, marc21_record):
+    """Test type to marc21 transformation."""
```
Comment (iGormilhit) [inline tests/unit/test_documents_dojson_marc21.py:259]:
```suggestion
    """Test type to MARC21 transformation."""
```

File: tests/unit/test_documents_dojson_marc21.py [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+            'b': _('docsubtype_atlas')
+        })
+    })
+    assert result == record
+
+
+def test_holdings_items_to_marc21(app, marc21_record, document,
+                                  holding_lib_martigny, item_lib_martigny,
+                                  ebook_5, holding_lib_martiny_electronic):
+    """Test holding items to marc21 transformation."""
```
Comment (iGormilhit) [inline tests/unit/test_documents_dojson_marc21.py:289]:
```suggestion
    """Test holding items to MARC21 transformation."""
```

**Review by iGormilhit — CHANGES_REQUESTED** [review]
Commit message proposition:

```
SRU: convert JSON to MARC21

* Adds dojson for JSON to MARC21 transformation.
```

File: rero_ils/modules/documents/dojson/contrib/jsontomarc21/model.py [hunk 11]
```diff
@@ -0,0 +1,507 @@
...
+                    'name': get_name(Organisation, hit.organisation.pid)
+                },
+                'library': {
+                    'pid': hit.library.pid,
+                    'name': get_name(Library, hit.library.pid)
+                },
+                'location': {
+                    'pid': hit.location.pid,
+                    'name': get_name(Location, hit.location.pid)
+                },
```
Comment (zannkukai) [inline rero_ils/modules/documents/dojson/contrib/jsontomarc21/model.py:122]: For each holding hit, you reload the organisation, the library, the location. But these informations should be load once and temporary store to improve speed. Something like :
```python
orgs = libraries = locations = {}
for hit in hits:
    holding = hit.to_dict()
    if hit.organisation.pid not in orgs:
        orgs[hit.organisation.pid] = get_name(Organisation, hit.organisation.pid)
    ...
    result = {
        'organisation': { 'pid': hit.organisation.pid, 'name': orgs[hit.organisation.pid]},
         ...
     }
     
```

File: rero_ils/modules/documents/serializers.py [hunk 12]
```diff
@@ -166,8 +174,300 @@ def _process_library_buckets(cls, org, lib_buckets):
...
+            ``do()`` function.
+        :param xslt_filename: XSLT filename. (Default: ``None``)
+        :param schema_class: The schema class. (Default: ``None``)
+        """
+        # xslt_filename = resource_filename(
+        #     'rero_ils', 'xslts/MARC21slim2MODS3-6.xsl')
+        if xslt_filename:
+            self.dumps_kwargs = dict(xslt_filename=xslt_filename)
+        else:
+            self.dumps_kwargs = {}
```
Comment (zannkukai) [inline rero_ils/modules/documents/serializers.py:197]: maybe better :
```python
self.dump_kwargs = {}
if xslt_filename:
    self.dumps_kwargs['xslt_filename'] = xslt_filename```

File: rero_ils/modules/documents/serializers.py [hunk 13]
```diff
@@ -166,8 +174,300 @@ def _process_library_buckets(cls, org, lib_buckets):
...
+
+        self.schema_class = schema_class
+        super().__init__()
+
+    def transform_record(self, pid, record, language,
+                         links_factory=None, **kwargs):
+        """Transform record into an intermediate representation."""
+        record = to_marc21.do(record, language=language,
+                              with_holdings_items=True)
+        return record
```
Comment (zannkukai) [inline rero_ils/modules/documents/serializers.py:207]: coud be merge into singe instruction : `return to_marc21.do(...)`

File: rero_ils/modules/documents/serializers.py [hunk 14]
```diff
@@ -166,8 +174,300 @@ def _process_library_buckets(cls, org, lib_buckets):
...
+
+    def transform_search_hit(self, pid, record_hit, language,
+                             links_factory=None, **kwargs):
+        """Transform search result hit into an intermediate representation."""
+        record = to_marc21.do(record_hit, language=language,
+                              with_holdings_items=True)
+        return record
```
Comment (zannkukai) [inline rero_ils/modules/documents/serializers.py:214]: same here

**Review by zannkukai — CHANGES_REQUESTED** [review]

**Review by iGormilhit — APPROVED** [review]
Commit message approved.

**Pull request merged**